### ISSUE TITLE
Update ACE Editor

### ISSUE THREAD

**Issue by MrSerth** [openHPI/codeocean#250]
Right now, an intermediate version of ACE between 1.1.8 and 1.1.9 is used. The release has been removed due to errors in the build script.
This is the commit we use: https://github.com/ajaxorg/ace-builds/tree/b14d255f5135bbcf814ec46fad35171e39ad1d82

```shell
$ diff -rq vendor/assets/javascripts/ace/ ace-builds-b14d255f5135bbcf814ec46fad35171e39ad1d82/src-min

Only in vendor/assets/javascripts/ace/: .DS_Store
Files vendor/assets/javascripts/ace/ace.js and ace-builds-b14d255f5135bbcf814ec46fad35171e39ad1d82/src-min/ace.js differ
```

**As a result, only `ace.js` needs to be checked, all other files are unchanged.**

<hr>

I guess, the `ace.js` is from another commit:

``` shell
$ for REV in $(git rev-list master); do
    git checkout $REV;
    git diff -w --shortstat --no-index src/ace.js <...>/CodeOcean/vendor/assets/javascripts/ace/ace.js; 
done

<snip>
HEAD is now at beb9ff68 package 03.03.15
 1 file changed, 112 insertions(+), 60 deletions(-)
Previous HEAD position was beb9ff68 package 03.03.15
HEAD is now at b14d255f package 24.02.15
 1 file changed, 109 insertions(+), 57 deletions(-)
Previous HEAD position was b14d255f package 24.02.15
HEAD is now at a4e495d8 package 20.12.14
 1 file changed, 57 insertions(+), 23 deletions(-)
Previous HEAD position was a4e495d8 package 20.12.14
HEAD is now at 7fafd126 package 08.11.14
 1 file changed, 363 insertions(+), 235 deletions(-)
Previous HEAD position was 7fafd126 package 08.11.14
HEAD is now at b3d5a61b package 09.10.14
 1 file changed, 510 insertions(+), 340 deletions(-)
Previous HEAD position was b3d5a61b package 09.10.14
</snap>
```

Due to that, I would use this commit: https://github.com/ajaxorg/ace-builds/blob/a4e495d8901876c6bafe3870a35cb8e32c827e97/src/ace.js

**Comment by MrSerth** [comment]
The only difference (between a few white space changes) is the addition of these lines at the very end:

``` js
            (function() {
                window.require(["ace/ace"], function(a) {
                    a && a.config.init(true);
                    if (!window.ace)
                        window.ace = a;
                    for (var key in a) if (a.hasOwnProperty(key))
                        window.ace[key] = a[key];
                });
            })();
```

Other changes are made through CSS.

**Comment by MrSerth** [comment]
We should remove the special UTF-16 handling when upgrading ACE: 2b11415d

**Comment by MrSerth** [comment]
These are the potential steps when working on this feature:

- [ ] (Re)move the anonymous function added to the ACE source files. Either, we could be able to remove it completely, or we could move it to another location, so that it is part of our own code
- [ ] Find a solution for the dynamically loaded ACE files, such as syntax parser or themes. A first step might be to add them via yarn and require them dynamically (requires a fixed URL, without pinning).
- [ ] Switch the main ACE editor to be loaded via yarn
- [ ] Finally update the ACE version; previous steps might still use an old version of ACE.

**Comment by MrSerth** [comment]
Progress report as of September 2023: We successfully moved our ACE version to a slightly more current release 🥳 .

As part of Kira's master thesis, we started working on pair programming support for CodeOcean. Of course, this feature requires a synchronized editor, which we added through #1843 (and later refined with further PRs). During testing, however, we identified a major roadblock of the current ACE editor: The `change` events triggered by each modification in the ACE editor. In our build, `change` events were triggered quite frequently and a single key press could issue multiple events. Specifically, an insert would trigger 1-2 events (first line + all remaining lines), and a remove 1-3 (last line + all lines in-between + first line) events. This made synchronizing the editor rather difficult, so that we decided to upgrade: Coming from something like 1.1.8, the next minor release 1.2.0 offered a reworked integration with `change` events, which would only trigger one event for each modification. This is what we looked for. Following a successful upgrade, we later went to the last release of the 1.2.x series.

During our update, we learned the following:
- Our ACE editor previously used was, indeed, between 1.1.8 and 1.1.9.
- We are using the so-called `src` build (not a minified version or a no conflict version [relating to the `require`]).
- The function snippet identified in [a previous comment](https://github.com/openHPI/codeocean/issues/250#issuecomment-578735774) is original ACE code.
- The ACE editor was modified to support custom annotations, which are using for the speech bubble icon when viewing comments on a RfC. This custom change was just restored with 1013903f.

As of today, we are using ACE 1.2.9 (downloaded from [here](https://github.com/ajaxorg/ace-builds/releases/tag/v1.2.9)). It is actively being used by learners in our current Python course with no complaint so far. The only observation we made is that scrolling in ACE is not working that smooth any longer.
